Add a Slider GUI component

Goldorion opened the pull request:
Hi

https://github.com/user-attachments/assets/d88f437f-a3b6-4c50-b132-53eca0004434


https://github.com/user-attachments/assets/114fbeda-44a0-4e06-8c66-9db84bbd239c

----

Edit by Klemen

Changelog:

* Added new custom GUI component types: slider
* Added new GUI procedure blocks: get slider value, set slider value

KlemenDEV commented:
Noticed I did not give you an official welcome back yet, so

> Welcome back 🥳

:D

KlemenDEV reviewed (changes requested):
Since we are dropping 1.21.4 support, please remove the files for this version. Thanks!

Update to 1.21.8 code is also needed.

KlemenDEV commented on src/main/java/net/mcreator/ui/dialogs/wysiwyg/SliderDialog.java:
```diff
@@ -0,0 +1,144 @@
+		sliderMachineName.enableRealtimeValidation();
+
+		JTextField sliderPrefix = new JTextField(8);
+		JSpinner minValueSpinner = new JSpinner(new SpinnerNumberModel(0, -10000.0, 10000.0, 1));
```
For min/max selection, use JMinMaxSpinner

I would also allow values from min to max float, not just to +/- 10000, unless there is some limit in the game

Goldorion replied:
I'll change it to JMinMaxSpinner (didn't know it exist, my bad).

However, about allowing float values, users can already write any number between -10k and +10k, as shown with the .0 at the end of each. I put the step value to 1 as I thought it would be the most common value used by users, but everything else in the code is made to support a double value.

Goldorion replied:
I can confirm that using decimal values inside the dialog works fine and cause no problem at all in the software or in the game. Mojang supports double values (and so NeoForge's `ExtendedSlider` class too) and this works perfectly with the current design.

KlemenDEV replied:
I did not mean that you should not use decimal values. I just meant that instead of hardcoded 10000 min max we can allow min and max float value instead

Goldorion replied:
Oh ok, I didn't understand that. Understood

Goldorion replied:
Yeah, I'll use the Double.MAX and MIM static variables. I didn't understand you were talking of them.

Goldorion replied:
Done

KlemenDEV reviewed (changes requested):
Checked a bit, I have one comment on the dialog.

AAli107 reviewed (changes requested):
Looks good so far, as I really like how customizable it is.

Though, I think it's better to have a procedure block just like the checkbox and text field where it gives the value of the GUI element, or you could make the "When slider moves" trigger work server side. You could maybe do both if possible.

<img width="591" height="98" alt="image" src="https://github.com/user-attachments/assets/e2a12928-2791-4451-b718-392676b97982" />

KlemenDEV reviewed (changes requested):
Looks good overall, but I have a systemic change suggestion/request

Goldorion commented:
Oh and the When Slider moves suggestion is already implemented btw

KlemenDEV commented:
Ali and me had same suggestion at the same time :D

Goldorion commented:
> Ali and me had same suggestion at the same time :D

But I don't think it's the easiest way. Sliders are far more similar to buttons than text fields/check boxes inside the code.

Goldorion commented on src/main/java/net/mcreator/element/parts/gui/Slider.java:
```diff
@@ -0,0 +1,69 @@
+	public double step;
+	public final String prefix;
+	public final String suffix;
+	public final Procedure whenSliderMoves;
```
Technically, slot input should follow this same logic, but they don't. Also, I don't think the form-like GUI is the most used case of GUI. Sliders are meant for situations like you have a block (or an item, whateverr it is) with a special behaviour and you want to have a value customizable, like the Jigsaw block is... and the jigsaw block use the same approach as I did. They execute their code when the slider moves. They don't try to read its value inside somewhere else.

KlemenDEV commented:
Sliders indeed extend buttons, but do we need change events that often? Tehcnically we should then also add them to text fields and checkboxes, as all of those are input form elements and one could argue one would need change event there just as much.

Button is final action of the form usually, not the sliding movement.

I think to keep the system uniformized, we should for now just add procedures and then add later change events if needed.

Change events are also heavier code-wise since they add new network message type, while just sending data state to server re-uses existing system we have for other UI components

AAli107 commented:
tbh, any kind of input field needs to be accessible in some way on the server side, or else they would turn out to be quite useless.

So a procedure block that will get the value of a specific slider is quite important.

Goldorion commented:
> tbh, any kind of input field needs to be accessible in some way on the server side, or else they would turn out to be quite useless.
> 
> So a procedure block that will get the value of a specific slider is quite important.

If you check carefully the video, you can see I show that I can use the slider with an explosion, so this proves I already made the slider component server-side usable. You don't have to worry about this. I had the same thought when I implemented it, so I added the packet (for the procedure trigger) directly at the beginning, so they are full of possibilities :D

Goldorion commented:
> The thing is most users don't even know how to access custom dependency of trigger to obtain value, while get/set are more understandable to the begginers.

For this argument, I can only say you're right. I'm already doing the changes to make it use MessageStateUpdate (which I still don't understand why it's used, but I know how, so I guess it's fine lol)

Goldorion commented:
Both getter and setter have been added too.

Goldorion commented:
That's weird because my test setup was this way:

When i click on a button: Set slider value to [Get slider value + 1]

And it didn't cause me a problem. I always had the following number (slider minimum was 0 and maximal was 10).

Goldorion commented:
Thinking to that, your explanations make it seem it returns the value when sliding it (so the vakue added/removed), but that's even more weird.

AAli107 commented:
> Thinking to that, your explanations make it seem it returns the value when sliding it (so the vakue added/removed), but that's even more weird.
> 
> Edit: what MC version are you on?

neoforge 1.21.8

KlemenDEV commented:
Could you share your test workspace maybe, it could help @Goldorion

AAli107 commented:
Here is my test workspace: [test.zip](https://github.com/user-attachments/files/22061178/test.zip)

Goldorion commented:
Just putting as a draft as I noticed major issues while re-testing.

Goldorion commented:
Ok, I don't know why it worked for my specific case, but I could confirm the problem. I was sending the corrected value for the procedures, but when I was sending the value for the procedure blocks (MenuStateUpdate) I was simply sending the value variable (which for some reasons is changed for a value between 0 and 1), causing your problem. It's fixed now... with an additionnal fix for the When slider moves packet.

AAli107 reviewed (changes requested):
Checks:
- Setting slider value ranges is as expected (value ranges between min and max you set)✅
- Default slider value is as expected (should match whatever the user sets)✅
- Getting slider value ranges is as expected ❌
  - default value is returned initially as expected until the slider is moved, where it only returns a number ranging between 0-1, where 0 is min, and 1 is max instead of the range the user set for the slider.

TLDR; Fix the getter inconsistency where it returns a range between 0 and 1 instead of a number between min and max of the slider.

Test workspace (slight change from last shared workspace): [test.zip](https://github.com/user-attachments/files/22061679/test.zip)

Goldorion commented:
I'm not sure to understand what you mean, because when I try to execute a procedure to print only the getter's value when the slider is moved, I obtain what is displayed on the slider as shown with this video. Can you explain a bit more your problem? Maybe I don't understand (and did you make sure to regenerate the code of the entire workspace before retesting?)

https://github.com/user-attachments/assets/f490aad3-cf8f-4877-8416-cdedf4487243

Goldorion commented:
I just tested with your test workspace and this works exactly like it should work, so I would really appreciate more explanations.

https://github.com/user-attachments/assets/d6467559-3bc8-48bd-ae8b-8621ef7e202d

AAli107 commented:
> I'm not sure to understand what you mean, because when I try to execute a procedure to print only the getter's value when the slider is moved, I obtain what is displayed on the slider as shown with this video. Can you explain a bit more your problem? Maybe I don't understand (and did you make sure to regenerate the code of the entire workspace before retesting?)
> 2025-08-30.20-18-41.mp4
> 
> This is the executed procedure when the slider is moved: <img alt="image" width="787" height="230" src="https://private-user-images.githubusercontent.com/38427877/483916948-c8b79a9e-ce33-4075-ad7a-59bc5ff5d24d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY2MjUzODEsIm5iZiI6MTc1NjYyNTA4MSwicGF0aCI6Ii8zODQyNzg3Ny80ODM5MTY5NDgtYzhiNzlhOWUtY2UzMy00MDc1LWFkN2EtNTliYzVmZjVkMjRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODMxVDA3MjQ0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk4MGUxNmI1ZGQ2NDgyNGUyMjI3OTQxMThkMTM1N2QzMmFlODc1N2I0YmQ3NDFlOGZhMzEwNTYyMWFiYmUzYmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.exMQUshX6J_JOIajMMLOFq1Z0695ujgS0GV_BT-60h8">

Does it show the same value if you use the "Send message to player" procedure block? Also make sure you are using neoforge 1.21.8 as to what I am testing.

My test procedure:
<img width="1297" height="358" alt="image" src="https://github.com/user-attachments/assets/60e439d3-5211-4f41-b7c3-cbce5d7591f6" />

KlemenDEV reviewed (changes requested):
I suggest one change for a bit cleaner and shorter code. While slider bases of button, I think since we use separate message IDs should be separate for sliders. Also one UI suggestion.

I have also tested workspace from @AAli107 and could not reproduce the problem, reported value goes correctly from -5 to 5 in my tests.

AAli107 reviewed (approved):
<img width="2560" height="1377" alt="2025-08-31_19 25 53" src="https://github.com/user-attachments/assets/fc36303d-ff4a-40a0-9523-0317df6ca44d" />


Ohh yeah, it works now. Everything looks good now. 👍

Goldorion commented on src/main/java/net/mcreator/ui/dialogs/wysiwyg/SliderDialog.java:
I implemented the real-time constraint as it was much easier and faster (and smaller in terms of changes) to make than developing a whole new thing that would work for both JSPinner at the same time.

Goldorion commented on plugins/generator-1.21.1/neoforge-1.21.1/templates/gui/gui_window.java.ftl:
```diff
@@ -305,6 +327,29 @@ public class ${name}Screen extends AbstractContainerScreen<${name}Menu> implemen
+			${component.getName()} = new ExtendedSlider(this.leftPos + ${component.gx(data.width)}, this.topPos + ${component.gy(data.height)},
+				${component.getWidth(w.getWorkspace())}, ${component.getHeight(w.getWorkspace())}, Component.translatable(
+				"gui.${modid}.${registryname}.${component.getName()}_prefix"), Component.translatable("gui.${modid}.${registryname}.${component.getName()}_suffix"),
+				${component.min}, ${component.max}, ${component.value}, ${component.step}, 0, true) {
```
Oh I think I know how it's used. I'll do a quick test, but I think it's just meant for the displayed value.

KlemenDEV replied:
Interesting, I will do some testing myself too just to be sure

KlemenDEV replied:
You are right, that is funny :D

KlemenDEV commented on src/main/java/net/mcreator/ui/dialogs/wysiwyg/SliderDialog.java:
Still can create situation like this one: 
<img width="258" height="111" alt="image" src="https://github.com/user-attachments/assets/3ce58bd6-9c63-400f-ac85-1a2446178b9d" />

This is why I hinted post-validation after OK is clicked may be easier than real-time handling :D

KlemenDEV replied:
Post-validation is actually much much easier, you just check for value and show dialog if it is not right, or simply don't save it. Other dialogs do it.

Anyway, I forgot steps to reproduce:

1. Set min max
2. Set current value
3. Increase min above current value

Goldorion replied:
> Post-validation is actually much much easier, you just check for value

Yes, but it requires a new Validator class for the value spinner + a new validator for the step size (because we don't use the same validation condition) + a new VJSpinner class to allow spinners to have a validation process attached to them.

Spectrall368 reviewed (approved):
it seems ok, however I don't know if the slider value should be saved or not when reopening the gui

KlemenDEV commented:
Don't think so, text fields or check boxes don't save value either
